In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as exp
import plotly.graph_objects as go
import zipfile

In [ ]:
zip = zipfile.ZipFile('archive.zip')

In [ ]:
df = pd.read_csv(zip.open('spi.csv'))

In [ ]:
df.head()

spi_rank      country  spi_score  basic_human_needs  wellbeing  \
0       1.0       Norway      92.63              95.29      93.30   
1       2.0      Finland      92.26              95.62      93.09   
2       3.0      Denmark      92.15              95.30      92.74   
3       4.0      Iceland      91.78              96.66      93.65   
4       5.0  Switzerland      91.78              95.25      93.80   

   opportunity  basic_nutri_med_care  water_sanitation  shelter  \
0        89.30                 98.81             98.33    93.75   
1        88.07                 98.99             99.26    96.48   
2        88.41                 98.62             98.21    94.92   
3        85.04                 98.99             98.82    93.16   
4        86.28                 98.72             98.96    92.97   

   personal_safety  access_basic_knowledge  access_info_comm  health_wellness  \
0            90.29                   98.66             95.80            89.32   
1            87.75                   96.32             95.14            85.73   
2            89.46                   97.44             98.18            85.15   
3            95.66                   99.51             93.12            91.02   
4            90.35                   98.60             95.07            91.50   

   env_quality  personal_rights  personal_freedom_choice  inclusiveness  \
0        89.44            96.34                    91.16          83.77   
1        95.15            96.13                    88.10          82.81   
2        90.20            97.08                    90.03          81.64   
3        90.93            95.14                    88.01          77.63   
4        90.05            96.69                    90.65          74.81   

   access_adv_edu  
0           85.92  
1           85.23  
2           84.89  
3           79.39  
4           82.99

In [ ]:
df.shape

(169, 18)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   spi_rank                 168 non-null    float64
 1   country                  169 non-null    object 
 2   spi_score                169 non-null    float64
 3   basic_human_needs        169 non-null    float64
 4   wellbeing                169 non-null    float64
 5   opportunity              169 non-null    float64
 6   basic_nutri_med_care     169 non-null    float64
 7   water_sanitation         169 non-null    float64
 8   shelter                  169 non-null    float64
 9   personal_safety          169 non-null    float64
 10  access_basic_knowledge   169 non-null    float64
 11  access_info_comm         169 non-null    float64
 12  health_wellness          169 non-null    float64
 13  env_quality              169 non-null    float64
 14  personal_rights          1

In [ ]:
df.isnull().sum()

spi_rank                   1
country                    0
spi_score                  0
basic_human_needs          0
wellbeing                  0
opportunity                0
basic_nutri_med_care       0
water_sanitation           0
shelter                    0
personal_safety            0
access_basic_knowledge     0
access_info_comm           0
health_wellness            0
env_quality                0
personal_rights            0
personal_freedom_choice    0
inclusiveness              0
access_adv_edu             0
dtype: int64

In [ ]:
df[df['spi_rank'].isnull()]

spi_rank country  spi_score  basic_human_needs  wellbeing  opportunity  \
168       NaN   World      65.05              74.18      64.42        56.54   

     basic_nutri_med_care  water_sanitation  shelter  personal_safety  \
168                 84.92             69.99    80.63             61.2   

     access_basic_knowledge  access_info_comm  health_wellness  env_quality  \
168                   72.03             70.22            60.18        55.27   

     personal_rights  personal_freedom_choice  inclusiveness  access_adv_edu  
168            60.16                    62.22          42.22           61.58

In [ ]:
df[df['country']=='World']

spi_rank country  spi_score  basic_human_needs  wellbeing  opportunity  \
168       NaN   World      65.05              74.18      64.42        56.54   

     basic_nutri_med_care  water_sanitation  shelter  personal_safety  \
168                 84.92             69.99    80.63             61.2   

     access_basic_knowledge  access_info_comm  health_wellness  env_quality  \
168                   72.03             70.22            60.18        55.27   

     personal_rights  personal_freedom_choice  inclusiveness  access_adv_edu  
168            60.16                    62.22          42.22           61.58

In [ ]:
df.dropna(inplace=True)

In [ ]:
df.shape

(168, 18)

In [ ]:
df.reset_index(inplace=True)

In [ ]:
df.head()

index  spi_rank      country  spi_score  basic_human_needs  wellbeing  \
0      0       1.0       Norway      92.63              95.29      93.30   
1      1       2.0      Finland      92.26              95.62      93.09   
2      2       3.0      Denmark      92.15              95.30      92.74   
3      3       4.0      Iceland      91.78              96.66      93.65   
4      4       5.0  Switzerland      91.78              95.25      93.80   

   opportunity  basic_nutri_med_care  water_sanitation  shelter  \
0        89.30                 98.81             98.33    93.75   
1        88.07                 98.99             99.26    96.48   
2        88.41                 98.62             98.21    94.92   
3        85.04                 98.99             98.82    93.16   
4        86.28                 98.72             98.96    92.97   

   personal_safety  access_basic_knowledge  access_info_comm  health_wellness  \
0            90.29                   98.66             95.80            89.32   
1            87.75                   96.32             95.14            85.73   
2            89.46                   97.44             98.18            85.15   
3            95.66                   99.51             93.12            91.02   
4            90.35                   98.60             95.07            91.50   

   env_quality  personal_rights  personal_freedom_choice  inclusiveness  \
0        89.44            96.34                    91.16          83.77   
1        95.15            96.13                    88.10          82.81   
2        90.20            97.08                    90.03          81.64   
3        90.93            95.14                    88.01          77.63   
4        90.05            96.69                    90.65          74.81   

   access_adv_edu  
0           85.92  
1           85.23  
2           84.89  
3           79.39  
4           82.99

In [ ]:
fig = exp.bar(df, x='country', y='spi_score')
fig.update_layout(title='SPI Score for every country', xaxis_title='Country', yaxis_title='SPI_Score', height=400, width=1000, showlegend=True)
fig.show()

Norway has the highest SPI score with the score of 92.63 followed by Iceland, Swedan and Germany with score of 92.26, 92.15, 91.78 respectively

In [ ]:
df.head()

index  spi_rank      country  spi_score  basic_human_needs  wellbeing  \
0      0       1.0       Norway      92.63              95.29      93.30   
1      1       2.0      Finland      92.26              95.62      93.09   
2      2       3.0      Denmark      92.15              95.30      92.74   
3      3       4.0      Iceland      91.78              96.66      93.65   
4      4       5.0  Switzerland      91.78              95.25      93.80   

   opportunity  basic_nutri_med_care  water_sanitation  shelter  \
0        89.30                 98.81             98.33    93.75   
1        88.07                 98.99             99.26    96.48   
2        88.41                 98.62             98.21    94.92   
3        85.04                 98.99             98.82    93.16   
4        86.28                 98.72             98.96    92.97   

   personal_safety  access_basic_knowledge  access_info_comm  health_wellness  \
0            90.29                   98.66             95.80            89.32   
1            87.75                   96.32             95.14            85.73   
2            89.46                   97.44             98.18            85.15   
3            95.66                   99.51             93.12            91.02   
4            90.35                   98.60             95.07            91.50   

   env_quality  personal_rights  personal_freedom_choice  inclusiveness  \
0        89.44            96.34                    91.16          83.77   
1        95.15            96.13                    88.10          82.81   
2        90.20            97.08                    90.03          81.64   
3        90.93            95.14                    88.01          77.63   
4        90.05            96.69                    90.65          74.81   

   access_adv_edu  
0           85.92  
1           85.23  
2           84.89  
3           79.39  
4           82.99

In [ ]:
legend_groups=['Basic Human Needs', 'Basic Nutrition and Medical Care']

In [ ]:
fig = go.Figure(layout=go.Layout(height=600, width=800, title=go.layout.Title(text="Basic Needs of Every Country")))
fig.add_trace(go.Scatter(x=df['country'], y=df['basic_human_needs'], hovertext=['country', 'basic_human_needs'], name=legend_groups[0]))
fig.add_trace(go.Scatter(x=df['country'], y=df['basic_nutri_med_care'], hovertext=['country', 'basic_nutri_med_care'], name=legend_groups[1]))
fig.update_layout(xaxis_title='Country', yaxis_title='Daily Needs')
fig.show()

Above Analysis shows the Basic Needs available for People of Every country. Norway is Leading again with the Top Score

In [ ]:
fig = exp.scatter(df, x="country", y="wellbeing", color="country",
                 title="Quality of Life in every Country")

fig.add_trace(
    go.Scatter(
        x=df['country'],
        y=df['opportunity'],
        hovertext=[df['country'], df['opportunity']],
        showlegend=True)
)
fig.update()
fig.show()

Above Analysis id done for Each and every countries. Lets look into Top countries with highest SPI Score.

In [ ]:
water = df[df['water_sanitation']>90]
water.sort_values(by='water_sanitation', ascending=False)

index  spi_rank      country  spi_score  basic_human_needs  wellbeing  \
25      25      26.0     Slovenia      85.83              93.58      86.21   
1        1       2.0      Finland      92.26              95.62      93.09   
35      35      36.0      Croatia      82.82              91.32      85.01   
4        4       5.0  Switzerland      91.78              95.25      93.80   
53      53      54.0      Belarus      74.83              86.72      78.05   
..     ...       ...          ...        ...                ...        ...   
118    118     119.0        Libya      58.23              75.73      55.08   
54      54      55.0      Armenia      74.56              86.80      73.87   
68      68      69.0     Paraguay      71.49              82.80      72.99   
40      40      41.0    Argentina      80.38              85.00      80.01   
50      50      51.0     Malaysia      75.22              87.23      78.33   

     opportunity  basic_nutri_med_care  water_sanitation  shelter  \
25         77.70                 97.59             99.27    89.14   
1          88.07                 98.99             99.26    96.48   
35         72.14                 97.20             99.06    93.39   
4          86.28                 98.72             98.96    92.97   
53         59.71                 98.08             98.95    92.33   
..           ...                   ...               ...      ...   
118        43.89                 90.76             91.58    76.44   
54         62.99                 94.95             91.21    91.09   
68         58.68                 91.54             91.17    83.55   
40         76.12                 94.31             90.34    88.91   
50         60.11                 91.33             90.29    91.84   

     personal_safety  access_basic_knowledge  access_info_comm  \
25             88.32                   95.38             82.18   
1              87.75                   96.32             95.14   
35             75.63                   96.31             85.94   
4              90.35                   98.60             95.07   
53             57.53                   96.17             73.60   
..               ...                     ...               ...   
118            44.14                   60.29             45.85   
54             69.96                   89.69             78.97   
68             64.93                   71.45             81.47   
40             66.45                   84.06             87.91   
50             75.47                   89.64             83.56   

     health_wellness  env_quality  personal_rights  personal_freedom_choice  \
25             81.25        86.04            90.06                    81.67   
1              85.73        95.15            96.13                    88.10   
35             77.45        80.34            88.66                    74.17   
4              91.50        90.05            96.69                    90.65   
53             67.69        74.73            42.09                    77.15   
..               ...          ...              ...                      ...   
118            58.68        55.48            35.19                    49.40   
54             66.08        60.75            80.60                    59.29   
68             61.49        77.54            81.10                    61.78   
40             69.03        79.03            91.59                    67.30   
50             67.64        72.50            66.22                    69.66   

     inclusiveness  access_adv_edu  
25           61.40           77.66  
1            82.81           85.23  
35           52.05           73.69  
4            74.81           82.99  
53           52.85           66.78  
..             ...             ...  
118          39.47           51.51  
54           47.95           64.14  
68           41.01           50.83  
40           68.70           76.89  
50           43.29           61.26  

[72 rows x 19 columns]

In [ ]:
fig = exp.scatter(df.query("spi_score >= 80"), 
                 x="water_sanitation", 
                 y="spi_score",
                 size="spi_score", 
                 color="country",
                 hover_name="country",
                 title= "Countries with Better Water Sanitation"
                )
fig.show()

Above Analysis shows the Water Sanitation of Top Country having SPI score of above 90. Top 5 countries with better water sanitation are slovenia with score 99.27, Finland with score 99.26, Croatia with score 99.06, Switzerland and Belarus with score 98.96 

In [ ]:
plt.Figure(figsize=(16, 7))
fig = exp.bar(df.query('spi_score>80'), x='country', y='personal_safety',title='Safety Rate in Country having SPI score above 80', hover_data=['country', 'personal_safety'])
fig.update_layout(xaxis_title='country', yaxis_title='Personal Safety', )
fig.show()

In [ ]:
fig = exp.imshow(df, x=df.columns, y=df.index, text_auto=True)
fig.update_layout(width=500,height=500)
fig.show()

In [ ]:
df.columns

Index(['index', 'spi_rank', 'country', 'spi_score', 'basic_human_needs',
       'wellbeing', 'opportunity', 'basic_nutri_med_care', 'water_sanitation',
       'shelter', 'personal_safety', 'access_basic_knowledge',
       'access_info_comm', 'health_wellness', 'env_quality', 'personal_rights',
       'personal_freedom_choice', 'inclusiveness', 'access_adv_edu'],
      dtype='object')

In [ ]:
inclusivness = df[['country', 'inclusiveness', 'spi_rank']].sort_values(by='inclusiveness', ascending=False)
inclusivness

country  inclusiveness  spi_rank
0          Norway          83.77       1.0
1         Finland          82.81       2.0
2         Denmark          81.64       3.0
6          Sweden          79.84       7.0
5          Canada          79.54       6.0
..            ...            ...       ...
167   South Sudan          13.42     168.0
104  Saudi Arabia          10.49     105.0
162         Yemen           8.65     163.0
95        Bahrain           8.12      96.0
131         Syria           4.26     132.0

[168 rows x 3 columns]

In [ ]:
country = inclusivness['country']

In [ ]:
fig = exp.choropleth(inclusivness, locations='country', color='inclusiveness', locationmode='country names')
fig.update_layout(title='Inclusiveness of every country', geo_scope='world', geo=dict(showframe=False, showcoastlines=False))
fig.show()

Norway and Finland has the highest Inclusiveness rate which is above 80.

In [ ]:
from plotly.offline import iplot

In [ ]:
values = dict(type = 'choropleth', 
           locations = df['country'],
           locationmode = 'country names',
           colorscale='Blues',
           z = df['spi_score'], 
           text = df['country'],
           colorbar = {'title':'Social Progress Index'})

layout = dict(title = 'Social Progress Index', 
              geo = dict(showframe = True, 
                         projection = {'type': 'azimuthal equal area'}))

figure = go.Figure(data = [values], layout=layout)
iplot(figure)